In [1]:
# Step 1: Import Libraries
import os
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from tqdm import tqdm

# Instantiate the LabelEncoder
label_encoder = LabelEncoder()

In [2]:
# Define paths for the datasets based on the current working directory
base_dir = os.path.join(os.getcwd())
behavior_file = os.path.join(base_dir,'behaviors.parquet')
history_file = os.path.join(base_dir,  'history.parquet')
articles_file = os.path.join(base_dir, 'articles.parquet')


Read the files

In [3]:
# Step 3: Load Datasets
# Load each dataset from the specified paths
behavior_df = pd.read_parquet(behavior_file)
# history_df = pd.read_parquet(history_file)
articles_df = pd.read_parquet(articles_file)

# Display the first few rows of each dataset to confirm loading worked
print("Behavior Data:")
display(behavior_df.head())

#print("\nHistory Data:")
#display(history_df.head())

print("\nArticles Data:")
display(articles_df.head())


Behavior Data:


impression_id  article_id     impression_time  read_time  \
0          47727         NaN 2023-05-21 21:35:07       20.0   
1          47731         NaN 2023-05-21 21:32:33       13.0   
2          47736         NaN 2023-05-21 21:33:32       17.0   
3          47737         NaN 2023-05-21 21:38:17       27.0   
4          47740         NaN 2023-05-21 21:36:02       48.0   

   scroll_percentage  device_type  \
0                NaN            1   
1                NaN            1   
2                NaN            1   
3                NaN            1   
4                NaN            1   

                                  article_ids_inview article_ids_clicked  \
0  [9482380, 9775183, 9744403, 9775297, 9774020, ...           [9775183]   
1      [9774557, 9774516, 9775331, 9775277, 9759966]           [9759966]   
2  [9759966, 9774557, 9775352, 9746360, 9772601, ...           [9774652]   
3  [9774580, 9775131, 9775202, 9774789, 9774972, ...           [9775184]   
4  [9774826, 9775171, 9775076, 9769624, 9775056, ...           [9774648]   

   user_id  is_sso_user  gender  postcode  age  is_subscriber  session_id  \
0    18293        False     NaN       NaN  NaN          False         265   
1    18293        False     NaN       NaN  NaN          False         265   
2    18293        False     NaN       NaN  NaN          False         265   
3    18293        False     NaN       NaN  NaN          False         265   
4    18293        False     NaN       NaN  NaN          False         265   

   next_read_time  next_scroll_percentage  
0            34.0                   100.0  
1            45.0                   100.0  
2            78.0                   100.0  
3             6.0                    52.0  
4            32.0                   100.0


Articles Data:


article_id                                title  \
0     3000022       Hanks beskyldt for mishandling   
1     3000063      Bostrups aske spredt i Furesøen   
2     3000613  Jesper Olsen ramt af hjerneblødning   
3     3000700             Madonna topløs med heste   
4     3000840              Otto Brandenburg er død   

                                            subtitle  last_modified_time  \
0  Tom Hanks har angiveligt mishandlet sin afdøde... 2023-06-29 06:20:32   
1       Studieværten blev mindet med glad festlighed 2023-06-29 06:20:32   
2  Den tidligere danske landsholdsspiller i fodbo... 2023-06-29 06:20:33   
3  47-årige Madonna poserer både topløs og sammen... 2023-06-29 06:20:33   
4  Sangeren og skuespilleren Otto Brandenburg er ... 2023-06-29 06:20:33   

   premium                                               body  \
0    False  Tom Hanks skulle angiveligt have mishandlet si...   
1    False  Strålende sensommersol. Jazzede toner. Glas me...   
2    False  Jesper Olsen, der er noteret for 43 kampe på d...   
3    False  Skal du have stillet Madonna-sulten inden konc...   
4    False  'Og lidt for Susanne, Birgitte og Hanne... ' '...   

       published_time           image_ids     article_type  \
0 2006-09-20 09:24:18           [3518381]  article_default   
1 2006-09-24 07:45:30  [3170935, 3170939]  article_default   
2 2006-05-09 11:29:00           [3164998]  article_default   
3 2006-05-04 11:03:12           [3172046]  article_default   
4 2007-03-01 18:34:00           [3914446]  article_default   

                                                 url  ...  \
0  https://ekstrabladet.dk/underholdning/udlandke...  ...   
1  https://ekstrabladet.dk/nyheder/samfund/articl...  ...   
2  https://ekstrabladet.dk/sport/fodbold/landshol...  ...   
3  https://ekstrabladet.dk/underholdning/udlandke...  ...   
4  https://ekstrabladet.dk/nyheder/samfund/articl...  ...   

          entity_groups                                             topics  \
0                 [PER]  [Kriminalitet, Kendt, Underholdning, Personfar...   
1                    []  [Kendt, Underholdning, Begivenhed, Personlig b...   
2  [LOC, PER, PER, PER]  [Kendt, Sport, Fodbold, Sundhed, Sygdom og beh...   
3                    []                   [Kendt, Livsstil, Underholdning]   
4                    []  [Kendt, Underholdning, Begivenhed, Personlig b...   

  category  subcategory   category_str total_inviews  total_pageviews  \
0      414        [432]  underholdning           NaN              NaN   
1      118        [133]        nyheder           NaN              NaN   
2      142   [196, 271]          sport           NaN              NaN   
3      414        [432]  underholdning           NaN              NaN   
4      118        [133]        nyheder           NaN              NaN   

   total_read_time  sentiment_score  sentiment_label  
0              NaN           0.9911         Negative  
1              NaN           0.5155          Neutral  
2              NaN           0.9876         Negative  
3              NaN           0.8786          Neutral  
4              NaN           0.9468         Negative  

[5 rows x 21 columns]

### Check for missing values

In [4]:
# Check for missing values
print(behavior_df.isnull().sum())
#print(history_df.isnull().sum())
print(articles_df.isnull().sum())

impression_id                    0
article_id                 8458027
impression_time                  0
read_time                        0
scroll_percentage          8523902
device_type                      0
article_ids_inview               0
article_ids_clicked              0
user_id                          0
is_sso_user                      0
gender                    11176907
postcode                  11795148
age                       11694243
is_subscriber                    0
session_id                       0
next_read_time              319928
next_scroll_percentage     1360982
dtype: int64
article_id                 0
title                      0
subtitle                   0
last_modified_time         0
premium                    0
body                       0
published_time             0
image_ids              14322
article_type               0
url                        0
ner_clusters               0
entity_groups              0
topics                     0
category       

In [5]:
print(behavior_df.columns)

Index(['impression_id', 'article_id', 'impression_time', 'read_time',
       'scroll_percentage', 'device_type', 'article_ids_inview',
       'article_ids_clicked', 'user_id', 'is_sso_user', 'gender', 'postcode',
       'age', 'is_subscriber', 'session_id', 'next_read_time',
       'next_scroll_percentage'],
      dtype='object')


We observed that we have a lot of missing values for the gender , age and postcode attributes and since they are not relevant for our NRMS we drop them.

In [6]:
# Now drop columns with high missing values
behavior_df = behavior_df.drop(columns=['impression_id', 'article_id','gender', 'postcode', 'age','scroll_percentage','device_type'])



In [7]:
print(behavior_df.isnull().sum())

impression_time                 0
read_time                       0
article_ids_inview              0
article_ids_clicked             0
user_id                         0
is_sso_user                     0
is_subscriber                   0
session_id                      0
next_read_time             319928
next_scroll_percentage    1360982
dtype: int64


In [8]:
# Now drop columns with high missing values
articles_df = articles_df.drop(columns=['image_ids'])


In [9]:
print(articles_df.columns)

Index(['article_id', 'title', 'subtitle', 'last_modified_time', 'premium',
       'body', 'published_time', 'article_type', 'url', 'ner_clusters',
       'entity_groups', 'topics', 'category', 'subcategory', 'category_str',
       'total_inviews', 'total_pageviews', 'total_read_time',
       'sentiment_score', 'sentiment_label'],
      dtype='object')


In [10]:
print(articles_df.isnull().sum())

article_id                 0
title                      0
subtitle                   0
last_modified_time         0
premium                    0
body                       0
published_time             0
article_type               0
url                        0
ner_clusters               0
entity_groups              0
topics                     0
category                   0
subcategory                0
category_str               0
total_inviews         107205
total_pageviews       108600
total_read_time       108600
sentiment_score            0
sentiment_label            0
dtype: int64


In [11]:
print(articles_df.columns)


Index(['article_id', 'title', 'subtitle', 'last_modified_time', 'premium',
       'body', 'published_time', 'article_type', 'url', 'ner_clusters',
       'entity_groups', 'topics', 'category', 'subcategory', 'category_str',
       'total_inviews', 'total_pageviews', 'total_read_time',
       'sentiment_score', 'sentiment_label'],
      dtype='object')


# Modife the article dataset

This code loads only the specified columns, converts subcategory_ids to a single string format, and applies LabelEncoder for model compatibility.

In [12]:
# Now that we have the correct column names, let's proceed with the correct script

import pandas as pd
from sklearn.preprocessing import LabelEncoder

# Specify the correct columns based on the user's provided column names
articles_columns = ['article_id', 'title', 'subtitle', 'last_modified_time', 'premium',
       'body', 'published_time', 'article_type', 'ner_clusters',
       'entity_groups', 'topics', 'category', 'subcategory', 'category_str',
       'total_inviews', 'total_pageviews', 'total_read_time',
       'sentiment_score', 'sentiment_label']

# Load only the necessary columns from the articles file
articles_df = pd.read_parquet(articles_file, columns=articles_columns)

# Convert subcategory from list to a single string (assuming it's a list of IDs)
articles_df['subcategory'] = articles_df['subcategory'].apply(lambda x: ' '.join(map(str, x)))

# Apply LabelEncoder to transform 'subcategory' into a numeric format
label_encoder = LabelEncoder()
articles_df['subcategory_encoded'] = label_encoder.fit_transform(articles_df['subcategory'])

# Display the resulting DataFrame to verify changes
articles_df[['article_id', 'title', 'subtitle', 'last_modified_time', 'premium',
       'body', 'published_time', 'article_type', 'ner_clusters',
       'entity_groups', 'topics', 'category','subcategory_encoded', 'category_str',
       'total_inviews', 'total_pageviews', 'total_read_time',
       'sentiment_score', 'sentiment_label']].head()



article_id                                title  \
0     3000022       Hanks beskyldt for mishandling   
1     3000063      Bostrups aske spredt i Furesøen   
2     3000613  Jesper Olsen ramt af hjerneblødning   
3     3000700             Madonna topløs med heste   
4     3000840              Otto Brandenburg er død   

                                            subtitle  last_modified_time  \
0  Tom Hanks har angiveligt mishandlet sin afdøde... 2023-06-29 06:20:32   
1       Studieværten blev mindet med glad festlighed 2023-06-29 06:20:32   
2  Den tidligere danske landsholdsspiller i fodbo... 2023-06-29 06:20:33   
3  47-årige Madonna poserer både topløs og sammen... 2023-06-29 06:20:33   
4  Sangeren og skuespilleren Otto Brandenburg er ... 2023-06-29 06:20:33   

   premium                                               body  \
0    False  Tom Hanks skulle angiveligt have mishandlet si...   
1    False  Strålende sensommersol. Jazzede toner. Glas me...   
2    False  Jesper Olsen, der er noteret for 43 kampe på d...   
3    False  Skal du have stillet Madonna-sulten inden konc...   
4    False  'Og lidt for Susanne, Birgitte og Hanne... ' '...   

       published_time     article_type  \
0 2006-09-20 09:24:18  article_default   
1 2006-09-24 07:45:30  article_default   
2 2006-05-09 11:29:00  article_default   
3 2006-05-04 11:03:12  article_default   
4 2007-03-01 18:34:00  article_default   

                                        ner_clusters         entity_groups  \
0                                    [David Gardner]                 [PER]   
1                                                 []                    []   
2  [Frankrig, Jesper Olsen, Jesper Olsen, Jesper ...  [LOC, PER, PER, PER]   
3                                                 []                    []   
4                                                 []                    []   

                                              topics  category  \
0  [Kriminalitet, Kendt, Underholdning, Personfar...       414   
1  [Kendt, Underholdning, Begivenhed, Personlig b...       118   
2  [Kendt, Sport, Fodbold, Sundhed, Sygdom og beh...       142   
3                   [Kendt, Livsstil, Underholdning]       414   
4  [Kendt, Underholdning, Begivenhed, Personlig b...       118   

   subcategory_encoded   category_str  total_inviews  total_pageviews  \
0                  219  underholdning            NaN              NaN   
1                   10        nyheder            NaN              NaN   
2                  101          sport            NaN              NaN   
3                  219  underholdning            NaN              NaN   
4                   10        nyheder            NaN              NaN   

   total_read_time  sentiment_score sentiment_label  
0              NaN           0.9911        Negative  
1              NaN           0.5155         Neutral  
2              NaN           0.9876        Negative  
3              NaN           0.8786         Neutral  
4              NaN           0.9468        Negative

This code will:

Convert last_modified_time and published_time to milliseconds.
Calculate the time_interval in milliseconds between the mod_time and pub_time.

In [13]:
# Converting last_modified_time and published_time to milliseconds and calculating the interval

# Ensure the timestamps are in datetime format
articles_df['last_modified_time'] = pd.to_datetime(articles_df['last_modified_time'])
articles_df['published_time'] = pd.to_datetime(articles_df['published_time'])

# Calculate mod_time and pub_time in milliseconds
articles_df['mod_time'] = articles_df['last_modified_time'].astype('int64') // 10**6  # Convert to milliseconds
articles_df['pub_time'] = articles_df['published_time'].astype('int64') // 10**6  # Convert to milliseconds

# Calculate the time interval between last_modified_time and published_time in milliseconds
articles_df['time_interval'] = articles_df['mod_time'] - articles_df['pub_time']

# Display the resulting DataFrame to verify changes
print(articles_df.head())


   article_id                                title  \
0     3000022       Hanks beskyldt for mishandling   
1     3000063      Bostrups aske spredt i Furesøen   
2     3000613  Jesper Olsen ramt af hjerneblødning   
3     3000700             Madonna topløs med heste   
4     3000840              Otto Brandenburg er død   

                                            subtitle  last_modified_time  \
0  Tom Hanks har angiveligt mishandlet sin afdøde... 2023-06-29 06:20:32   
1       Studieværten blev mindet med glad festlighed 2023-06-29 06:20:32   
2  Den tidligere danske landsholdsspiller i fodbo... 2023-06-29 06:20:33   
3  47-årige Madonna poserer både topløs og sammen... 2023-06-29 06:20:33   
4  Sangeren og skuespilleren Otto Brandenburg er ... 2023-06-29 06:20:33   

   premium                                               body  \
0    False  Tom Hanks skulle angiveligt have mishandlet si...   
1    False  Strålende sensommersol. Jazzede toner. Glas me...   
2    False  Jesper Olse

In [14]:
print(articles_df.columns)

Index(['article_id', 'title', 'subtitle', 'last_modified_time', 'premium',
       'body', 'published_time', 'article_type', 'ner_clusters',
       'entity_groups', 'topics', 'category', 'subcategory', 'category_str',
       'total_inviews', 'total_pageviews', 'total_read_time',
       'sentiment_score', 'sentiment_label', 'subcategory_encoded', 'mod_time',
       'pub_time', 'time_interval'],
      dtype='object')


### Modify the behavior dataset

In [15]:
print(behavior_df.columns)

Index(['impression_time', 'read_time', 'article_ids_inview',
       'article_ids_clicked', 'user_id', 'is_sso_user', 'is_subscriber',
       'session_id', 'next_read_time', 'next_scroll_percentage'],
      dtype='object')


 Add a feature (articles_num) for the count of in-view articles, and explode article_ids_inview to have one article per row for each impression.

In [16]:
import pandas as pd

# Step 1: Create a new DataFrame with only relevant columns and calculate articles_num for each row
behavior_df_new = behavior_df[['impression_time', 'article_ids_inview', 'user_id', 'session_id']]
behavior_df_new['articles_num'] = behavior_df['article_ids_inview'].apply(len)  # Calculate articles_num first

# Track the original and expected row counts
original_row_count = len(behavior_df_new)
expected_row_count = behavior_df_new['articles_num'].sum()  # Expected row count after explosion
print(f"Original row count: {original_row_count}")
print(f"Expected row count after explosion: {expected_row_count}")

# Step 2: Define the chunk size
chunk_size = 100000
num_chunks = len(behavior_df_new) // chunk_size + 1  # Total number of chunks

# Step 3: Process and save each chunk
exploded_total_row_count = 0  # Track total rows after explosion
for i, start in enumerate(range(0, len(behavior_df_new), chunk_size)):
    # Select a chunk of the data
    chunk = behavior_df_new.iloc[start:start + chunk_size].copy()  # Use .copy() to avoid SettingWithCopyWarning

    # Explode the chunk to have one article per row
    exploded_chunk = chunk.explode('article_ids_inview')

    # Track row count for each exploded chunk
    exploded_chunk_row_count = len(exploded_chunk)
    exploded_total_row_count += exploded_chunk_row_count
    print(f"Chunk {i+1}: Original rows = {len(chunk)}, Exploded rows = {exploded_chunk_row_count}")

    # Save the exploded chunk to disk with a unique filename
    exploded_chunk.to_parquet(f"exploded_behavior_data_chunk_{i+1}.parquet", index=False)

print(f"Total exploded row count after processing all chunks: {exploded_total_row_count}")
print("All chunks have been processed and saved to disk.")



<ipython-input-16-ae2704d6e866>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  behavior_df_new['articles_num'] = behavior_df['article_ids_inview'].apply(len)  # Calculate articles_num first


Original row count: 12063890
Expected row count after explosion: 133810641
Chunk 1: Original rows = 100000, Exploded rows = 1108995
Chunk 2: Original rows = 100000, Exploded rows = 1112646
Chunk 3: Original rows = 100000, Exploded rows = 1108097
Chunk 4: Original rows = 100000, Exploded rows = 1093400
Chunk 5: Original rows = 100000, Exploded rows = 1111439
Chunk 6: Original rows = 100000, Exploded rows = 1106701
Chunk 7: Original rows = 100000, Exploded rows = 1131349
Chunk 8: Original rows = 100000, Exploded rows = 1099479
Chunk 9: Original rows = 100000, Exploded rows = 1121967
Chunk 10: Original rows = 100000, Exploded rows = 1105815
Chunk 11: Original rows = 100000, Exploded rows = 1110836
Chunk 12: Original rows = 100000, Exploded rows = 1106868
Chunk 13: Original rows = 100000, Exploded rows = 1101245
Chunk 14: Original rows = 100000, Exploded rows = 1107548
Chunk 15: Original rows = 100000, Exploded rows = 1120077
Chunk 16: Original rows = 100000, Exploded rows = 1108779
Chunk 

In [17]:
import pyarrow.parquet as pq
import pyarrow as pa
import glob

# Define output file
output_file = "full_behavior_data.parquet"

# Initialize variables
first_chunk = True

# Initialize the Parquet writer only once, for the first chunk
for file in glob.glob("exploded_behavior_data_chunk_*.parquet"):
    # Read the chunk
    chunk = pq.read_table(file)

    # Write the first chunk to initialize the file with schema
    if first_chunk:
        # Open the Parquet writer with the schema of the first chunk
        writer = pq.ParquetWriter(output_file, chunk.schema)
        first_chunk = False

    # Write the current chunk to the output file
    writer.write_table(chunk)

# Close the writer after all chunks are written
writer.close()

print("All chunks have been merged into 'full_behavior_data.parquet'.")


All chunks have been merged into 'full_behavior_data.parquet'.


The above script merges all exploded chunks into a single parquet file, full behavior data .

In [18]:
behavior_exploded_file = os.path.join(base_dir, 'full_behavior_data.parquet')
behavior_exploded = pd.read_parquet(behavior_exploded_file)


In [19]:
# Print the number of rows
num_rows = len(behavior_exploded)
num_rows

133810641

In [20]:
# Check row count before explosion
original_rows = len(behavior_df)
print(f"Rows in original behavior_df: {original_rows}")

# Check row count after explosion
exploded_rows = len(behavior_exploded)
print(f"Rows in exploded behavior_df_new: {exploded_rows}")


Rows in original behavior_df: 12063890
Rows in exploded behavior_df_new: 133810641


The merged full_behavior_data.parquet file contains 133,810,641 rows. This confirms that all chunks were successfully processed and combined into the final dataset. Let me know if you need further analysis or specific processing on this dataset!

We read the full_behavior_data.parquet to behavior_df_new

In [21]:

# Alternatively, print the first 100 rows for a sample view:
print(behavior_df['article_ids_inview'].head(100))  # Shows first 100 entries


0     [9482380, 9775183, 9744403, 9775297, 9774020, ...
1         [9774557, 9774516, 9775331, 9775277, 9759966]
2     [9759966, 9774557, 9775352, 9746360, 9772601, ...
3     [9774580, 9775131, 9775202, 9774789, 9774972, ...
4     [9774826, 9775171, 9775076, 9769624, 9775056, ...
                            ...                        
95    [9767481, 9774840, 9774864, 9775042, 9775323, ...
96        [9775371, 9773947, 9775432, 9770028, 9771051]
97    [9775361, 9772706, 9770288, 9775419, 9775402, ...
98        [9759544, 9773947, 9771051, 9775371, 9774461]
99        [9759966, 9775352, 9774557, 9774652, 9775277]
Name: article_ids_inview, Length: 100, dtype: object


In [22]:
# Step-by-step code to calculate, sort, and display

# Calculate the number of articles in each list in 'article_ids_inview'
behavior_df['articles_num'] = behavior_df['article_ids_inview'].apply(len)

# Sort by 'articles_num' in ascending order
sorted_behavior_df = behavior_df.sort_values(by='articles_num')

# Display the first 100 rows of sorted results
print(sorted_behavior_df[['article_ids_inview', 'articles_num']].head(100))



                                     article_ids_inview  articles_num
8075203   [9514481, 9486486, 9135506, 9779186, 9521552]             5
8075205   [9777182, 9779285, 9779289, 9779269, 9778945]             5
9444306   [7213923, 9778788, 9778939, 9779541, 9779511]             5
9444310   [9772548, 9779507, 9777397, 9779007, 9779289]             5
9444314   [9779423, 9779705, 9779657, 9779045, 9779738]             5
...                                                 ...           ...
10085056  [9566633, 9759891, 9773364, 9773461, 9772434]             5
6107415   [9769624, 9287091, 9772869, 9772903, 9772813]             5
6107418   [9771859, 9772772, 9769800, 9772502, 9772923]             5
6107420   [9772923, 9772668, 9771859, 9772925, 9769800]             5
6107422   [9772772, 9772925, 9772502, 9771859, 9772923]             5

[100 rows x 2 columns]


In [23]:
# Step 1: Verify original row count and the minimum value in articles_num
original_row_count = len(behavior_df)
min_articles_num = behavior_df['articles_num'].min()
print(f"Original row count: {original_row_count}")
print(f"Minimum articles_num value: {min_articles_num}")



Original row count: 12063890
Minimum articles_num value: 5


In [24]:
print(behavior_exploded.columns)

Index(['impression_time', 'article_ids_inview', 'user_id', 'session_id',
       'articles_num'],
      dtype='object')


In [25]:
behavior_exploded.rename(columns={'article_ids_inview': 'article_id'}, inplace=True)


In [26]:
print(behavior_exploded.columns)

Index(['impression_time', 'article_id', 'user_id', 'session_id',
       'articles_num'],
      dtype='object')


Now we will create an impr_time that converts impression_time to an integer representation in milliseconds,

impr_pub_interval: calculates the difference between the impression and published times, giving the delay bewteem publishing and the user's impression

mpr_pub_hour that converts this interval into hours for easier groupin and analysis


1. Extract article_id from behavior_df_new ( after exploding article_ids_inview).
2. Merge behavior_df_new with articles_df on article_id.
3. Calculate impr_time, impr_pub_interval, and impr_pub_hour based on impression_time and published_time from the merged dataset.

In [27]:
import pandas as pd

# Ensure the necessary columns are in datetime format
# Convert both 'impression_time' and 'published_time' to datetime if not already done
behavior_exploded['impression_time'] = pd.to_datetime(behavior_exploded['impression_time'])
articles_df['published_time'] = pd.to_datetime(articles_df['published_time'])


# Merge behavior_df with articles_df on article_id to bring in published_time information
merged_df = behavior_exploded.merge(articles_df[['article_id', 'published_time']], on='article_id', how='left')

# 1. Convert impression_time to an integer representation in milliseconds
merged_df['impr_time'] = merged_df['impression_time'].astype('int64') // 10**6  # Convert from nanoseconds to milliseconds

# 2. Calculate the delay (in milliseconds) between the impression and the published time
merged_df['impr_pub_interval'] = (merged_df['impression_time'] - merged_df['published_time']).dt.total_seconds() * 1000

# 3. Convert the impr_pub_interval to hours
merged_df['impr_pub_hour'] = merged_df['impr_pub_interval'] / (1000 * 3600)  # Convert milliseconds to hours

# Display a sample to verify
print(merged_df[['impression_time', 'published_time', 'impr_time', 'impr_pub_interval', 'impr_pub_hour']].head())


      impression_time      published_time   impr_time  impr_pub_interval  \
0 2023-05-21 18:04:10 2023-05-21 17:20:39  1684692250          2611000.0   
1 2023-05-21 18:04:10 2023-05-21 16:34:14  1684692250          5396000.0   
2 2023-05-21 18:04:10 2023-05-21 11:52:46  1684692250         22284000.0   
3 2023-05-21 18:04:10 2023-05-21 16:02:24  1684692250          7306000.0   
4 2023-05-21 18:04:10 2023-05-21 17:58:57  1684692250           313000.0   

   impr_pub_hour  
0       0.725278  
1       1.498889  
2       6.190000  
3       2.029444  
4       0.086944  


Calculate impr_pub_interval:

impr_pub_interval computes the delay in milliseconds between when the article was published (published_time) and when the impression occurred (impression_time).
For example, if an article was published at 2023-11-13 08:00:00 and viewed at 2023-11-14 10:15:30, the impr_pub_interval might be 90090000 milliseconds (around 25 hours).


In [28]:
merged_df.head()

impression_time  article_id  user_id  session_id  articles_num  \
0 2023-05-21 18:04:10     9775184    60280    39582465            10   
1 2023-05-21 18:04:10     9775042    60280    39582465            10   
2 2023-05-21 18:04:10     9774532    60280    39582465            10   
3 2023-05-21 18:04:10     9774554    60280    39582465            10   
4 2023-05-21 18:04:10     9774826    60280    39582465            10   

       published_time   impr_time  impr_pub_interval  impr_pub_hour  
0 2023-05-21 17:20:39  1684692250          2611000.0       0.725278  
1 2023-05-21 16:34:14  1684692250          5396000.0       1.498889  
2 2023-05-21 11:52:46  1684692250         22284000.0       6.190000  
3 2023-05-21 16:02:24  1684692250          7306000.0       2.029444  
4 2023-05-21 17:58:57  1684692250           313000.0       0.086944

In [29]:
print(behavior_exploded.columns)

Index(['impression_time', 'article_id', 'user_id', 'session_id',
       'articles_num'],
      dtype='object')


In [30]:
print(merged_df.columns)

Index(['impression_time', 'article_id', 'user_id', 'session_id',
       'articles_num', 'published_time', 'impr_time', 'impr_pub_interval',
       'impr_pub_hour'],
      dtype='object')


In [31]:
print(articles_df.columns)

Index(['article_id', 'title', 'subtitle', 'last_modified_time', 'premium',
       'body', 'published_time', 'article_type', 'ner_clusters',
       'entity_groups', 'topics', 'category', 'subcategory', 'category_str',
       'total_inviews', 'total_pageviews', 'total_read_time',
       'sentiment_score', 'sentiment_label', 'subcategory_encoded', 'mod_time',
       'pub_time', 'time_interval'],
      dtype='object')


In [32]:
import pandas as pd

#Concatenate the specified columns from articles_df into merged_df based on article_id
merged_df = merged_df.merge(
     articles_df[['article_id', 'title', 'subtitle', 'last_modified_time', 'premium',
                 'body', 'published_time', 'article_type', 'ner_clusters', 'entity_groups',
                 'topics', 'category', 'subcategory', 'category_str', 'total_inviews',
                 'total_pageviews', 'total_read_time', 'sentiment_score', 'sentiment_label',
                 'subcategory_encoded', 'mod_time', 'pub_time', 'time_interval']],
    on='article_id',
    how='left'
)

# Display a sample of the updated merged_df to verify
print(merged_df.head())


      impression_time  article_id  user_id  session_id  articles_num  \
0 2023-05-21 18:04:10     9775184    60280    39582465            10   
1 2023-05-21 18:04:10     9775042    60280    39582465            10   
2 2023-05-21 18:04:10     9774532    60280    39582465            10   
3 2023-05-21 18:04:10     9774554    60280    39582465            10   
4 2023-05-21 18:04:10     9774826    60280    39582465            10   

     published_time_x   impr_time  impr_pub_interval  impr_pub_hour  \
0 2023-05-21 17:20:39  1684692250          2611000.0       0.725278   
1 2023-05-21 16:34:14  1684692250          5396000.0       1.498889   
2 2023-05-21 11:52:46  1684692250         22284000.0       6.190000   
3 2023-05-21 16:02:24  1684692250          7306000.0       2.029444   
4 2023-05-21 17:58:57  1684692250           313000.0       0.086944   

                                           title  ... category_str  \
0  Holger Rune taber til russer i Masters-finale  ...        sport   


Analyzing hourly differences in article impressions offers several insights, particularly for understanding engagement patterns and optimizing content strategy. Here’s how this data can be useful:

1. Identify Peak Engagement Times
By examining which hours show the largest increases or decreases in impressions, you can identify when articles are most engaging for readers.
For example, a sharp rise in the first few hours might indicate a strong initial interest, while steady impressions over a longer period could suggest enduring relevance.
2. Evaluate Article Popularity Trends
Articles with consistent growth in impressions may indicate ongoing interest, while those with declining impressions can signal waning attention.
Comparing these trends across different articles can help in identifying what types of content hold readers' interest longer.


In [33]:
print(merged_df.columns)


Index(['impression_time', 'article_id', 'user_id', 'session_id',
       'articles_num', 'published_time_x', 'impr_time', 'impr_pub_interval',
       'impr_pub_hour', 'title', 'subtitle', 'last_modified_time', 'premium',
       'body', 'published_time_y', 'article_type', 'ner_clusters',
       'entity_groups', 'topics', 'category', 'subcategory', 'category_str',
       'total_inviews', 'total_pageviews', 'total_read_time',
       'sentiment_score', 'sentiment_label', 'subcategory_encoded', 'mod_time',
       'pub_time', 'time_interval'],
      dtype='object')


In [34]:
print(behavior_exploded.columns)

Index(['impression_time', 'article_id', 'user_id', 'session_id',
       'articles_num'],
      dtype='object')


In [35]:
print(behavior_df.columns)

Index(['impression_time', 'read_time', 'article_ids_inview',
       'article_ids_clicked', 'user_id', 'is_sso_user', 'is_subscriber',
       'session_id', 'next_read_time', 'next_scroll_percentage',
       'articles_num'],
      dtype='object')


In [36]:
print(articles_df.columns)

Index(['article_id', 'title', 'subtitle', 'last_modified_time', 'premium',
       'body', 'published_time', 'article_type', 'ner_clusters',
       'entity_groups', 'topics', 'category', 'subcategory', 'category_str',
       'total_inviews', 'total_pageviews', 'total_read_time',
       'sentiment_score', 'sentiment_label', 'subcategory_encoded', 'mod_time',
       'pub_time', 'time_interval'],
      dtype='object')


In [37]:
print(merged_df.columns)

Index(['impression_time', 'article_id', 'user_id', 'session_id',
       'articles_num', 'published_time_x', 'impr_time', 'impr_pub_interval',
       'impr_pub_hour', 'title', 'subtitle', 'last_modified_time', 'premium',
       'body', 'published_time_y', 'article_type', 'ner_clusters',
       'entity_groups', 'topics', 'category', 'subcategory', 'category_str',
       'total_inviews', 'total_pageviews', 'total_read_time',
       'sentiment_score', 'sentiment_label', 'subcategory_encoded', 'mod_time',
       'pub_time', 'time_interval'],
      dtype='object')


In [38]:
# Print the number of rows in merged_df
num_rows = len(merged_df)
print(f"The number of rows in merged_df: {num_rows}")


The number of rows in merged_df: 133810641


In [39]:
merged_df.head()

impression_time  article_id  user_id  session_id  articles_num  \
0 2023-05-21 18:04:10     9775184    60280    39582465            10   
1 2023-05-21 18:04:10     9775042    60280    39582465            10   
2 2023-05-21 18:04:10     9774532    60280    39582465            10   
3 2023-05-21 18:04:10     9774554    60280    39582465            10   
4 2023-05-21 18:04:10     9774826    60280    39582465            10   

     published_time_x   impr_time  impr_pub_interval  impr_pub_hour  \
0 2023-05-21 17:20:39  1684692250          2611000.0       0.725278   
1 2023-05-21 16:34:14  1684692250          5396000.0       1.498889   
2 2023-05-21 11:52:46  1684692250         22284000.0       6.190000   
3 2023-05-21 16:02:24  1684692250          7306000.0       2.029444   
4 2023-05-21 17:58:57  1684692250           313000.0       0.086944   

                                           title  ... category_str  \
0  Holger Rune taber til russer i Masters-finale  ...        sport   
1           Har intet tilovers for FCK-udmelding  ...        sport   
2                        SLUT: Ekstase i Parken!  ...        sport   
3        Smed vigtige point: Kan skrive historie  ...        sport   
4    Vanvittige scener: FCK dukker AGF til sidst  ...        sport   

  total_inviews  total_pageviews total_read_time sentiment_score  \
0      203321.0          39827.0       2538445.0          0.7068   
1      188765.0          28656.0       2644095.0          0.6983   
2      559795.0         179557.0      20007134.0          0.5617   
3      174296.0          22331.0       1924591.0          0.8455   
4      313534.0          51251.0       5056938.0          0.9408   

  sentiment_label subcategory_encoded    mod_time    pub_time  time_interval  
0         Neutral                 209  1688021338  1684689639        3331699  
1        Negative                  43  1688021338  1684686854        3334484  
2        Positive                  43  1688021337  1684669966        3351371  
3        Positive                  43  1688021337  1684684944        3336393  
4         Neutral                  43  1688021337  1684691937        3329400  

[5 rows x 31 columns]

In [40]:
columns_to_drop = [
    'impression_time', 'articles_num', 'last_modified_time', 'body',
    'published_time_x', 'published_time_y', 'subcategory', 'ner_clusters',
    'entity_groups'
]

In [43]:
print(merged_df.columns)

Index(['impression_time', 'article_id', 'user_id', 'session_id',
       'articles_num', 'published_time_x', 'impr_time', 'impr_pub_interval',
       'impr_pub_hour', 'title', 'subtitle', 'last_modified_time', 'premium',
       'body', 'published_time_y', 'article_type', 'ner_clusters',
       'entity_groups', 'topics', 'category', 'subcategory', 'category_str',
       'total_inviews', 'total_pageviews', 'total_read_time',
       'sentiment_score', 'sentiment_label', 'subcategory_encoded', 'mod_time',
       'pub_time', 'time_interval'],
      dtype='object')


In [47]:
# Define columns to drop
columns_to_drop = [
    'impression_time', 'articles_num', 'last_modified_time', 'body',
    'published_time_x', 'published_time_y', 'subcategory', 'ner_clusters',
    'entity_groups','impr_pub_hour'
]

# Find missing columns
missing_columns = [col for col in columns_to_drop if col not in merged_df.columns]
print(f"Missing columns: {missing_columns}")


Missing columns: []


In [48]:
print("Columns in merged_df:")
print(list(merged_df.columns))

print("\nColumns to drop:")
print(columns_to_drop)


Columns in merged_df:
['impression_time', 'article_id', 'user_id', 'session_id', 'articles_num', 'published_time_x', 'impr_time', 'impr_pub_interval', 'impr_pub_hour', 'title', 'subtitle', 'last_modified_time', 'premium', 'body', 'published_time_y', 'article_type', 'ner_clusters', 'entity_groups', 'topics', 'category', 'subcategory', 'category_str', 'total_inviews', 'total_pageviews', 'total_read_time', 'sentiment_score', 'sentiment_label', 'subcategory_encoded', 'mod_time', 'pub_time', 'time_interval']

Columns to drop:
['impression_time', 'articles_num', 'last_modified_time', 'body', 'published_time_x', 'published_time_y', 'subcategory', 'ner_clusters', 'entity_groups', 'impr_pub_hour']


In [46]:
# Remove missing columns from the drop list
columns_to_drop = [col for col in columns_to_drop if col in merged_df.columns]

# Drop the remaining columns
final_df = merged_df.drop(columns=columns_to_drop)

# Verify the result
print(final_df.columns)


Index(['article_id', 'user_id', 'session_id', 'impr_time', 'impr_pub_interval',
       'impr_pub_hour', 'title', 'subtitle', 'premium', 'article_type',
       'topics', 'category', 'category_str', 'total_inviews',
       'total_pageviews', 'total_read_time', 'sentiment_score',
       'sentiment_label', 'subcategory_encoded', 'mod_time', 'pub_time',
       'time_interval'],
      dtype='object')


In [49]:
# Drop the remaining columns
final_df = final_df.drop(columns=['impr_pub_hour'])

In [52]:
# Verify the result
print(final_df.columns)

Index(['article_id', 'user_id', 'session_id', 'impr_time', 'impr_pub_interval',
       'title', 'subtitle', 'premium', 'article_type', 'topics', 'category',
       'category_str', 'total_inviews', 'total_pageviews', 'total_read_time',
       'sentiment_score', 'sentiment_label', 'subcategory_encoded', 'mod_time',
       'pub_time', 'time_interval'],
      dtype='object')


In [51]:
# Print the number of rows in merged_df
num_rows = len(final_df)
print(f"The number of rows in merged_df: {num_rows}")

The number of rows in merged_df: 133810641


In [53]:
import pyarrow.parquet as pq
import pyarrow as pa

# Define the output file
output_file = "final_dataset.parquet"

# Initialize the Parquet writer
table = pa.Table.from_pandas(final_df.iloc[:1])  # Use the first row to get schema
writer = pq.ParquetWriter(output_file, table.schema, compression='gzip')

# Define chunk size
chunk_size = 1_000_000  # Adjust based on memory availability
total_rows = len(final_df)

print(f"Starting to write {total_rows} rows in chunks...")

# Write DataFrame in chunks
for start in range(0, total_rows, chunk_size):
    chunk = final_df.iloc[start:start + chunk_size]
    table = pa.Table.from_pandas(chunk)
    writer.write_table(table)
    print(f"Written rows {start} to {start + len(chunk) - 1}.")

# Close the writer
writer.close()

print(f"Final dataset saved as '{output_file}'")

Starting to write 133810641 rows in chunks...
Written rows 0 to 999999.
Written rows 1000000 to 1999999.
Written rows 2000000 to 2999999.
Written rows 3000000 to 3999999.
Written rows 4000000 to 4999999.
Written rows 5000000 to 5999999.
Written rows 6000000 to 6999999.
Written rows 7000000 to 7999999.
Written rows 8000000 to 8999999.
Written rows 9000000 to 9999999.
Written rows 10000000 to 10999999.
Written rows 11000000 to 11999999.
Written rows 12000000 to 12999999.
Written rows 13000000 to 13999999.
Written rows 14000000 to 14999999.
Written rows 15000000 to 15999999.
Written rows 16000000 to 16999999.
Written rows 17000000 to 17999999.
Written rows 18000000 to 18999999.
Written rows 19000000 to 19999999.
Written rows 20000000 to 20999999.
Written rows 21000000 to 21999999.
Written rows 22000000 to 22999999.
Written rows 23000000 to 23999999.
Written rows 24000000 to 24999999.
Written rows 25000000 to 25999999.
Written rows 26000000 to 26999999.
Written rows 27000000 to 27999999.
W